### PR TITLE
Fix epoch logging on train epoch end

### LINKED ISSUE #12851: Same step maps to different epoch values with TensorBoardLogger and on_epoch=True
## 🐛 Bug

When using the `TensorBoardLogger` and the `on_epoch=True` option, it now happens in pl 1.6 that the same step number maps to different epoch values. This is probably due to [this PR](on_epoch=True).

### To Reproduce

```python
import os
import pprint
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.loggers import TensorBoardLogger
from tensorboard.backend.event_processing.event_accumulator import EventAccumulator


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss, on_epoch=True)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    logger = TensorBoardLogger(save_dir='.', name='', version='')

    model = BoringModel()
    trainer = Trainer(
        logger=logger,
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=3,
        enable_model_summary=False,
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)


if __name__ == "__main__":
    run()

    acc = EventAccumulator('.')
    acc.Reload()
    pprint.pprint(acc.Scalars('epoch'))
```

### Expected behavior

With pl 1.6.1 I get:
```python
[ScalarEvent(wall_time=1650628947.5834746, step=0, value=0.0),                                                                                                                                                                                                                                                                                                    
 ScalarEvent(wall_time=1650628947.5858712, step=0, value=1.0),
 ScalarEvent(wall_time=1650628947.589659, step=1, value=1.0),
 ScalarEvent(wall_time=1650628947.5918398, step=1, value=2.0),
 ScalarEvent(wall_time=1650628947.5954714, step=2, value=2.0),
 ScalarEvent(wall_time=1650628947.5977085, step=2, value=3.0)]
```

and with pl 1.5.10

```
[ScalarEvent(wall_time=1650629199.73182, step=0, value=0.0),                                                                                                                                                                                                                                                                                                      
 ScalarEvent(wall_time=1650629199.7335322, step=0, value=0.0),
 ScalarEvent(wall_time=1650629199.7365856, step=1, value=1.0),
 ScalarEvent(wall_time=1650629199.738278, step=1, value=1.0),
 ScalarEvent(wall_time=1650629199.7413125, step=2, value=2.0),
 ScalarEvent(wall_time=1650629199.744041, step=2, value=2.0)]
```

I do not know why every step is logged twice, but I guess there is a reason for it. However, I find it a bit weird that the same step now maps to different epoch values.

### Environment

- PyTorch Lightning Version (e.g., 1.5.0): 1.5.10 or 1.6.1
- PyTorch Version (e.g., 1.10): 1.11
- Python version (e.g., 3.9): 3.9
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: 11.3
- How you installed PyTorch (`conda`, `pip`, source): pip

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12851

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @otaj @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

### CODE REVIEW

File: tests/trainer/logging_/test_train_loop_logging.py [hunk 1]
```diff
@@ -31,7 +33,8 @@
 from tests.helpers.runif import RunIf
 
 
-def test__training_step__log(tmpdir):
+@mock.patch("pytorch_lightning.loggers.TensorBoardLogger.log_metrics")
+def test__training_step__log(mock_log_metrics, tmpdir):
```
Comment (carmocca) [inline tests/trainer/logging_/test_train_loop_logging.py:37]: Can you write a new test instead? This test is already way too broad.

Comment (rohitgr7) [reply tests/trainer/logging_/test_train_loop_logging.py:36]: done!

**Review by carmocca — APPROVED** [review]
Removed the for loop inside the test, I  think it's much easier to see this way and it's fine to unroll since it's not long

File: tests/trainer/logging_/test_train_loop_logging.py [hunk 2]
```diff
@@ -748,3 +750,37 @@ def validation_epoch_end(self, *_) -> None:
...
+        enable_checkpointing=False,
+        enable_progress_bar=False,
+    )
+    trainer.fit(model)
+
+    mock_log_metrics.assert_has_calls(
+        [
+            call(metrics={"foo_step": 0.0, "epoch": 0}, step=0),
+            call(metrics={"foo_step": 0.0, "epoch": 0}, step=1),
+            call(metrics={"foo_epoch": 0.0, "epoch": 0}, step=1),
```
Comment (carmocca) [inline tests/trainer/logging_/test_train_loop_logging.py:781]: For other reviewers, in master the `*_epoch` lines would have the `"epoch": N` value increased by 1

**Review by akihironitta — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
